Order Creation: Add discard changes prompt to new order customer details

rachelmcr opened the pull request:
Closes: #6454 

## Description

This adds a "discard changes" prompt when swiping to dismiss the Customer Details modal on the New Order screen, to prevent accidental data loss. The prompt only appears when there are unsaved changes.

The prompt is added with a SwiftUI view modifier, which should be reusable with other SwiftUI modal views in the future.

## Changes

* Adds a SwiftUI view modifier `discardChangesPrompt(canDismiss:didDismiss:)`:
   * This modifier wraps the SwiftUI in a UIHostingController (in `DiscardChangesWrapper`), which uses `canDismiss` to determine if the modal can be dismissed with a swipe gesture and fires `didDismiss` when the modal is dismissed.
   * If `canDismiss` is false, it uses our `UIAlertController.presentDiscardChangesActionSheet` helper to display the discard changes prompt.
* Moves `hasPendingChanges()` from the `EditOrderAddressFormViewModel` (used in order editing) to `AddressFormViewModelProtocol` so it can also be used for order creation to determine if the address form has unsaved changes.
* Adds a manually created binding `canDismiss` to `CreateOrderAddressFormViewModel`. I couldn't find a better way to create this binding to use with the new view modifier; let me know if you have an alternate suggestion for this.
* Adds the view modifier to the modal in `OrderCustomerSection`, using the new binding to determine if the modal can be dismissed without the prompt.

## Testing

1. Go to the Orders tab and create a new order.
2. Select "Add Customer Details."
3. Confirm you can close the modal using the "Close" button and swiping to dismiss it before you've made any changes.
4. Add some customer details and try swiping to dismiss the modal. Confirm a "discard changes" prompt appears and works as expected.
5. Add some customer details and confirm you can tap the "Done" button to save your changes.

## Screenshots


https://user-images.githubusercontent.com/8658164/160631566-7f8e2dff-814e-4e3a-b100-a17a2fca3687.mp4



## Submitter Checklist

Update release notes:
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

rachelmcr committed:
Add SwiftUI view modifier with discard changes prompt

rachelmcr committed:
Add discard changes prompt to customer details modal in order creation

Ecarrion commented on WooCommerce/Classes/View Modifiers/View+DiscardChanges.swift:
```diff
@@ -0,0 +1,70 @@
+        }
+
+        func presentationControllerDidAttemptToDismiss(_ presentationController: UIPresentationController) {
+            UIAlertController.presentDiscardChangesActionSheet(viewController: presentationController.presentedViewController) { [weak self] in
```
I wonder if we could inject somehow this action sheet, I could see cases where we don't want this specific message, what do you think?

Ecarrion reviewed (changes requested):
Hi @rachelmcr, 

Your `@Binding` question made me realize that I think we are interpreting the `SwiftUI` lifecycle in a slightly wrong way.

Basically, we need to treat the wrapper view as a disposable view, and update the underlying coordinator(state manager) on each update pass.

This implies that we need to hold the `addressFormViewModel` in some way, ideally in the `newOrderViewModel`.

As these changes are quite scattered(and I needed to validate what I'm saying 😛) I have used some rough changes for you to take a look at [in this diff](https://github.com/woocommerce/woocommerce-ios/compare/issue/6465-customer-details-discard-changes...issue/discard-proposed-changes?expand=1)

Let me know what you think!

rachelmcr commented:
@Ecarrion Thank you! I suspected I was missing something there, and that the `@Binding` wasn't quite right since we only needed to read the value, but I only got myself part of the way to the solution you worked out. (And partway there was never enough for it to work right, hehe.)

From a quick look at your changes I think this makes a lot of sense. 👍  I'll look at them more closely and work them into this PR, along with the other enhancement you suggested.

rachelmcr committed:
Fix lifecycle for address form view model

Co-authored-by: Ernesto Carrion <ernesto.carrion@automattic.com>

rachelmcr committed:
Inject action sheet for more flexibility with the prompt

rachelmcr committed:
Merge branch 'trunk' into issue/6465-customer-details-discard-changes

rachelmcr commented:
@Ecarrion This is ready for another look! I incorporated your proposed changes in [87aa140](https://github.com/woocommerce/woocommerce-ios/pull/6534/commits/87aa140794602a82ba8e4c483b0ab9f40c89250f), holding the `addressFormViewModel` in the `newOrderViewModel`. It is passed it from there to the `OrderCustomerSection`, where it is an observed object (to make sure the `hasPendingChanges` value is observed correctly). When the address form is closed, a trigger is fired to reset the view model to make sure changes are discarded.

I also followed your suggestion to inject the discard changes action sheet in [bc496ea](https://github.com/woocommerce/woocommerce-ios/pull/6534/commits/bc496ea54a6b3cb7bca54adc03805d66d09decb7), to make that view modifier more flexible.

Ecarrion commented on WooCommerce/Classes/View Modifiers/View+DiscardChanges.swift:
```diff
@@ -0,0 +1,70 @@
+
+    func makeUIViewController(context: Context) -> UIViewControllerType {
+        let viewController = UIHostingController(rootView: view)
+        return viewController
```
super nit 😅 
```suggestion
        UIHostingController(rootView: view)
```

Ecarrion commented on WooCommerce/Classes/View Modifiers/View+DiscardChanges.swift:
```diff
@@ -0,0 +1,70 @@
+        Coordinator(self)
+    }
+
+    class Coordinator: NSObject, UIAdaptivePresentationControllerDelegate {
```
```suggestion
    private class Coordinator: NSObject, UIAdaptivePresentationControllerDelegate {
```

rachelmcr replied:
This can't be made private because it's part of the `UIViewControllerRepresentable` protocol that `DiscardChangesWrapper` conforms to.

Ecarrion commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomerSection/OrderCustomerSection.swift:
```diff
@@ -7,22 +7,27 @@ struct OrderCustomerSection: View {
+                                      canDismiss: !addressFormViewModel.hasPendingChanges,
+                                      didDismiss: addressFormViewModel.userDidCancelFlow)
+                .onDisappear {
+                    viewModel.resetAddressFormTrigger.send()
```
Could you explain a little bit the reason about why you preferred a `Combine` publisher over of a regular method?

rachelmcr replied:
🤔  When I first set up that trigger, I thought the separate method `createOrderAddressFormViewModel()` wouldn't be needed and everything could be done within `configureAddressFormViewModel()` using `Combine`. That didn't work out, so now that I'm looking at this with fresh eyes I think this probably isn't needed and I could just simplify this line like this:

```
viewModel.addressFormViewModel = viewModel.createOrderAddressFormViewModel()
```

rachelmcr replied:
Updated in [7626e15](https://github.com/woocommerce/woocommerce-ios/pull/6534/commits/7626e15d3db238d837e898ae6fb089818f755e3d).

Ecarrion commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/NewOrder.swift:
```diff
@@ -98,7 +98,8 @@ struct NewOrder: View {
 
-                        OrderCustomerSection(viewModel: viewModel)
+                        OrderCustomerSection(viewModel: viewModel,
+                                             addressFormViewModel: viewModel.addressFormViewModel ?? viewModel.createOrderAddressFormViewModel())
```
Do you think we should wrap this inside some kind of `if let`? From looking at the code it seems like the  `viewModel.createOrderAddressFormViewModel()` part will never be needed but It's misleading to the reader.

rachelmcr replied:
Added this in [7626e15](https://github.com/woocommerce/woocommerce-ios/pull/6534/commits/7626e15d3db238d837e898ae6fb089818f755e3d), with documentation to clarify that it's just to handle the optional `viewModel.addressFormViewModel`.

Ecarrion commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomerSection/CreateOrderAddressFormViewModel.swift:
```diff
@@ -36,6 +37,15 @@ final class CreateOrderAddressFormViewModel: AddressFormViewModel, AddressFormVi
 
+    /// Creates and presents a discard changes action sheet.
+    ///
+    func presentDiscardChangesActionSheet(viewController: UIViewController) {
+        UIAlertController.presentDiscardChangesActionSheet(viewController: viewController) { [weak self] in
+            viewController.dismiss(animated: true, completion: nil)
+            self?.userDidCancelFlow()
+        }
+    }
```
Seeing the `ViewModel` having to `import UIKit` raised my alarm 😆 

I think we should keep the `UIKit` bit constrained to the `DiscardChangesWrapper`, what do you think about just allowing the consumer to provide the `title` and the `action button title`, so then we can call the UIKit action sheet inside the wrapper. something like:

```swift
func discardChangesPrompt(title: String,
                          discardButtonTitle: String,
                          canDismiss: Bool,
                          didDismiss: (() -> Void)? = nil) -> some View {
```

rachelmcr replied:
That's fair. I was trying to make it flexible enough to handle action sheets like the one we use for product details, which has more actions (discard changes, save as draft, cancel). But we don't have a need for that yet, and if we get to that point we can always handle the complexity in a `UIAlertController` helper rather than importing UIKit li.

I'll go ahead and change it to something more like your suggestion 👍

rachelmcr replied:
Updated in [59c090e](https://github.com/woocommerce/woocommerce-ios/pull/6534/commits/59c090eb868f19e7c4e382adfc6cae2fd99e0f5a). I decided to provide a set of default strings (since I expect that in most cases where we have a discard changes action sheet we'll be ok with those defaults) with the option to override them.

Ecarrion reviewed (commented):
Looking much better!
Left you another suggestion to keep our view models clean from the `UIKit` domain!

rachelmcr committed:
PR feedback

rachelmcr committed:
Simplify address form view model configuration

rachelmcr committed:
Create action sheet with option to override default strings

rachelmcr commented:
Thanks again for the feedback, @Ecarrion. This is ready for another look!

I'm still not entirely happy with the way the `CreateOrderAddressFormViewModel` view model is being configured (it feels a little hacky) but I couldn't think of a better way to do it, and at least this way is fairly simple/straightforward compared to earlier iterations.

Ecarrion commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomerSection/OrderCustomerSection.swift:
```diff
@@ -7,22 +7,26 @@ struct OrderCustomerSection: View {
+                .discardChangesPrompt(canDismiss: !addressFormViewModel.hasPendingChanges,
+                                      didDismiss: addressFormViewModel.userDidCancelFlow)
+                .onDisappear {
+                    viewModel.addressFormViewModel = viewModel.createOrderAddressFormViewModel()
```
One minor nit, I would have created a method called `viewModel.resetAddressForm()` to hide this complexity. Then we can add comments on why is this needed.

Ecarrion commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/NewOrder.swift:
```diff
@@ -131,6 +131,16 @@ struct NewOrder: View {
+
+    /// View model for Customer section address form.
+    ///
+    private var addressFormViewModel: CreateOrderAddressFormViewModel {
+        guard let addressFormViewModel = viewModel.addressFormViewModel else {
+            // Returns a newly created view model if one isn't set. This shouldn't be needed but handles the optional CreateOrderAddressFormViewModel.
+            return viewModel.createOrderAddressFormViewModel()
+        }
+        return addressFormViewModel
+    }
```
I wonder if this should leave inside the `vm`. what do you think?

Ecarrion reviewed (approved):
Looks great!, just some minor nits but then 🚀

rachelmcr committed:
Use address form view model method to reset address form on cancel

rachelmcr commented:
@Ecarrion Could I trouble you to take one more look at this? Your comments above made me rethink the approach — in particular, why we're resetting the address form by recreating the view model in the `NewOrderViewModel`. It occurs to me that a more straightforward approach is to reset it (by resetting all the address fields) from within `CreateOrderAddressFormViewModel` as part of the cancel action.

I tried that in [c776816](https://github.com/woocommerce/woocommerce-ios/pull/6534/commits/c776816f19d80408d5b7a2311f34008d8a4c6b09) and it simplifies a lot of things, and I think makes it clearer what we're doing. What do you think?

rachelmcr committed:
Update unit tests to set address form view model

Ecarrion commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/NewOrderViewModel.swift:
```diff
@@ -130,6 +130,19 @@ final class NewOrderViewModel: ObservableObject {
 
+    /// View model for the customer details address form.
+    ///
+    var addressFormViewModel: CreateOrderAddressFormViewModel {
+        CreateOrderAddressFormViewModel(siteID: siteID,
+                                        addressData: .init(billingAddress: orderSynchronizer.order.billingAddress,
+                                                           shippingAddress: orderSynchronizer.order.shippingAddress),
+                                        onAddressUpdate: { [weak self] updatedAddressData in
+            let input = Self.createAddressesInput(from: updatedAddressData)
+            self?.orderSynchronizer.setAddresses.send(input)
+            self?.trackCustomerDetailsAdded()
+        })
+    }
```
I tested the feature and it seems to work but I can see that now this view model is not being retained, right?

Should we make this a lazy var to be on the safe side?

rachelmcr replied:
Oh, you're right. But making this a lazy var doesn't work as-is because then what we pass in for its `addressData` isn't ever updated. Handling that takes us back to something more like how it was done in [7626e15](https://github.com/woocommerce/woocommerce-ios/pull/6534/commits/7626e15d3db238d837e898ae6fb089818f755e3d), so it seems I was mistaken about this being a better approach. 😬 

I think I'll revert these latest changes and address the comments from [your previous review](https://github.com/woocommerce/woocommerce-ios/pull/6534#pullrequestreview-932565888) + the UI issue you found here.

Ecarrion reviewed (commented):
Looks good, I just have one concern with the lifecycle of the `addressFormViewModel`.

Also, I think we messed with the landscape constraints when we added the wrapper.
<img width="884" alt="Screen Shot 2022-04-06 at 8 04 52 AM" src="https://user-images.githubusercontent.com/562080/161982170-9ebc7ad0-a321-42ac-a41a-c3b6c8ce1a84.png">

rachelmcr committed:
Revert "Update unit tests to set address form view model"

This reverts commit 8d2d4a1c1152c4a847c7bd5188f8a2abdc843a10.

rachelmcr committed:
Revert "Use address form view model method to reset address form on cancel"

This reverts commit c776816f19d80408d5b7a2311f34008d8a4c6b09.

rachelmcr committed:
Consolidate complexity of configuring and resetting address form view model

rachelmcr committed:
Remove extra safe area insets from DiscardChangesWrapper

rachelmcr commented:
@Ecarrion Ok, one more time. :) I reverted my latest changes to start from the approach used in [this previous review](https://github.com/woocommerce/woocommerce-ios/pull/6534#pullrequestreview-932565888). In [0296f26](https://github.com/woocommerce/woocommerce-ios/pull/6534/commits/0296f266594b5c2589c55136112c2e1782139958) I made a couple changes to it:

1. I created a `resetAddressForm` method that recreates the address form view model, and used that in `OrderCustomerSection`.
2. I made `NewOrderViewModel.addressFormViewModel` non-optional, to avoid that point of confusion. Instead, I initialize that property with a temporary view model first and then use `resetAddressForm` right away to reset it with the expected values. (I think this is less confusing than how it was done before, but let me know if you think it's better the other way.)

I also fixed the landscape issue — that was a weird one but it has a fairly simple fix, made in [992a409](https://github.com/woocommerce/woocommerce-ios/pull/6534/commits/992a409ed6d94f4d8f778d6433f245ebc44849e2).

Thanks for bearing with me while I wrap my head around the details of SwiftUI lifecycles!

Ecarrion reviewed (approved):
Looks & tests well! 🚀 

_PS: Sorry for dragging it for so long!_
